Fj 2024 lesson 8

gw3nax opened the pull request.

gw3nax committed:
solved task 7

gw3nax committed:
deleted some useless code

gw3nax committed:
Solved task 8

gw3nax committed:
Update ErrorResponseWrapper.java

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/controller/ErrorResponseWrapper.java:
```diff
@@ -0,0 +1,52 @@
+    @ExceptionHandler({CBRUnavailableException.class})
+    public ResponseEntity<ErrorResponse> handleCBRUnavailableException(CBRUnavailableException e) {
+        log.error("CBRUnavailableException was caught with reason: {}", e.getMessage());
+        var errorResponse = new ErrorResponse();
+        errorResponse.setMessage(e.getMessage());
+        errorResponse.setCode(500);
```
Стоит добавить builder и его использоваь

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/CurrencyExchangerApplication.java:
```diff
@@ -0,0 +1,17 @@
+package ru.gw3nax.currency_exchanger;
```
Не стоит использовать snake_case в названиях пакетов, они именуются в camelCase в Java

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/client/CBRClient.java:
```diff
@@ -0,0 +1,54 @@
+            return xmlMapper.readValue(entity, new TypeReference<>() {});
+        } catch (IOException e) {
+            throw new CBRUnavailableException("CBR service is unavailable")
+                    .addHeader("Retry-After", "3600");
```
Не стоит передавать в эксепшне заголовки, это не его ответственность.
Лучше определять эту логику в месте формирования ответа - т.е. в  @ExceptionHandler

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/controller/dto/ConvertCurrencyResponse.java:
```diff
@@ -0,0 +1,14 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class ConvertCurrencyResponse {
+    String fromCurrency;
+    String toCurrency;
+    Double convertedAmount;
```
Почему не private?

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/controller/dto/CurrencyResponse.java:
```diff
@@ -0,0 +1,13 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class CurrencyResponse {
+    String charCode;
+    Double value;
```
И тут

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/controller/dto/ConvertCurrencyRequest.java:
```diff
@@ -0,0 +1,14 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class ConvertCurrencyRequest {
+    String fromCurrency;
+    String toCurrency;
+    Double amount;
```
И тут

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/service/CurrencyExchangeService.java:
```diff
@@ -0,0 +1,97 @@
+        this.cbrClient = cbrClient;
+    }
+
+    private static void validateRequest(ConvertCurrencyRequest convertCurrencyRequest) {
+        validateCharCode(convertCurrencyRequest.getToCurrency());
+        validateCharCode(convertCurrencyRequest.getFromCurrency());
+        if (convertCurrencyRequest.getAmount() < 0) {
+            throw new IllegalRequestArgumentsException("Amount must be greater than zero");
+        }
+    }
+
+    public static void validateCharCode(String charCode) {
+        try {
+            Currency.getInstance(charCode);
+        } catch (IllegalArgumentException e) {
+            throw new NotExistingCharCodeException("Currency does not exist.");
+        }
+    }
```
Наивная валидация это тоже вариант, но в следующих этапах стоит использовать Jakarta Bean Validation или другие системные подходы

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/service/CurrencyExchangeService.java:
```diff
@@ -0,0 +1,97 @@
+        CurrencyResponse toCurrencyRate;
+        double convertedAmount;
+
+        if (convertCurrencyRequest.getFromCurrency().equals(RUB_CODE)) {
```
При сравнении с константой стоит инвертировать порядок, что не ловить NPE
RUB_CODE.equals(...)

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/service/CurrencyExchangeService.java:
```diff
@@ -0,0 +1,97 @@
+                    convertCurrencyRequest,
+                    new BigDecimal(convertedAmount).setScale(2, RoundingMode.HALF_UP).doubleValue()
+            );
+        } else if (convertCurrencyRequest.getToCurrency().equals(RUB_CODE)) {
```
Сейм

engarik commented on currency_exchanger/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,25 @@
+spring:
```
Если делать отступы между разными блоками конфиг проще читать

engarik commented on currency_exchanger/src/test/java/ru/gw3nax/currency_exchanger/CurrencyResponseExchangerApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class CurrencyResponseExchangerApplicationTests {
```
Дефолтный тест, можно удалять

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/client/CBRClient.java:
```diff
@@ -0,0 +1,54 @@
+@Slf4j
+public class CBRClient {
+
+    String BASE_URL;
+    private final XmlMapper xmlMapper = new XmlMapper();
+    private HttpClient httpClient = HttpClientBuilder.create().build();
+
+    @Autowired
+    CBRClient(ApplicationConfig applicationConfig) {
+        this.BASE_URL = applicationConfig.cbrBaseUrl();
+    }
```
Плохой подход. У тебя нарушена инкапсуляция только ради тестов.
Настрой клиент так, чтобы он читал url из конфига, тогда настройка для тестов и основного кода будет одинаковой

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/controller/ErrorResponseWrapper.java:
```diff
@@ -0,0 +1,52 @@
+    @ExceptionHandler({NotFoundException.class})
+    public ResponseEntity<ErrorResponse> handleNotFoundException(NotFoundException e) {
+        log.error("NotFoundException was caught with reason: {}", e.getMessage());
+        return ResponseEntity.status(404).body(new ErrorResponse(404, e.getMessage()));
```
У ResponseEntity есть билдер notFound()

gw3nax replied:
У билдера notFound() ведь нет возможности добавить body к ResponseEntity

engarik replied:
Ок

engarik commented on currency_exchanger/src/main/java/ru/gw3nax/currency_exchanger/controller/ErrorResponseWrapper.java:
```diff
@@ -0,0 +1,52 @@
+        for (Map.Entry<String, String> entry : e.getHeaders().entrySet()) {
+            header.add(entry.getKey(), entry.getValue());
+        }
+       return new ResponseEntity<>(errorResponse, header, HttpStatus.valueOf(500));
```
500 - это internal server error, тебе нужен 503

engarik reviewed (commented):
Привет!
13/15
-1 за реализацию клиента
-1 за неверный response для кейса с недоступностью валют

gw3nax committed:
fix
set BASEURL field to private final
remove headers from exception
edit CBRClientTest
set all fields in dto to private
add builder to response dto's

engarik reviewed (commented):
Принято ✅
